okio/jvm/src/main/java/okio/Options.kt

```diff
@@ -31,7 +32,209 @@ class Options private constructor(
   companion object {
     @JvmStatic
     fun of(vararg byteStrings: ByteString): Options {
-      return Options(byteStrings.clone()) // Defensive copy.
+      // Sort the byte strings which is required when to recursively build the trie. Map the sorted
+      // indexes to the caller's indexes.
+      val sortedByteStrings = byteStrings.sortedArray()
+      // TODO(jwilson): this is a O(N*N) loop to recover the original indexes. Ick.
+      val sortedIndexes = IntArray(sortedByteStrings.size) {
+        byteStrings.indexOf(sortedByteStrings[it])
+      }
+
+      require(sortedByteStrings.isNotEmpty()) { "no options provided" }
+      require(sortedByteStrings[0].size() > 0) { "the empty byte string is not a supported option" }
+
+      // Strip elements that will never be returned because they follow their own prefixes. For
+      // example, if the caller provides ["abc", "abcde"] we will never return "abcde" because we
+      // return as soon as we encounter "abc".
+      val trieByteStrings = mutableListOf<ByteString>()
+      val trieIndexes = mutableListOf<Int>()
+      collectEligible@
+      for (i in 0 until sortedByteStrings.size) {
+        val byteString = sortedByteStrings[i]
+        val index = sortedIndexes[i]
+        for (j in trieIndexes.size - 1 downTo 0) {
+          if (byteString.startsWith(trieByteStrings[j]) && index >= trieIndexes[j]) {
+            continue@collectEligible // This value can never be returned.
+          }
+        }
+        trieByteStrings += byteString
+        trieIndexes += index
+      }
+
+      val trieBytes = Buffer()
+      buildTrieRecursive(
+          node = trieBytes,
+          byteStrings = trieByteStrings,
+          indexes = trieIndexes)
+
+      val trie = IntArray(trieBytes.intCount.toInt())
+      var i = 0
+      while (!trieBytes.exhausted()) {
+        trie[i++] = trieBytes.readInt()
+      }
+
+      return Options(byteStrings.clone() /* Defensive copy. */, trie)
+    }
+
+    /**
+     * Builds a trie encoded as an int array. Nodes in the trie are of two types: SELECT and SCAN.
```

Comment:
A picture is worth 1000 words. Picture here:
https://github.com/square/okio/pull/432/files#diff-a2a8b008b6caa5abb9b320fc3e9e38deR70